expan/core/experiment.py

```diff
@@ -290,20 +295,63 @@ def _quantile_filtering(self, data, kpis, percentile, threshold_type):
         
         :param kpis: the kpis to perform filtering
         :type  kpis: list[str]
-        :param percentile: percentile considered as filtering threshold
-        :type  percentile: float
-        :param threshold_type: type of threshold used ('lower' or 'upper')
-        :type  threshold_type: str
+        :param thresholds: dict of thresholds mapping KPI names to (type, percentile) tuples
+        :type  thresholds: dict
 
         :return: boolean values indicating whether the row should be filtered
         :rtype: pd.Series
         """
-        method_table   = {'upper': lambda x: x > threshold, 'lower': lambda x: x <= threshold}
-        na_replacement = {'upper': -float_info.max,         'lower':  float_info.max}
 
-        for column in data[kpis].columns:
-            threshold = np.percentile(data[column].fillna(na_replacement[threshold_type]), percentile)
-            flags = flags | data[column].apply(method_table[threshold_type])
+        def find_smallest(data, quantile):
+            """ Return boolean vector of data points smaller than given quantile."""
+            threshold = data.quantile(quantile)
+            return data.apply(lambda x: x < threshold)
+
+        def find_largest(data, quantile):
+            """ Return boolean vector of data points larger than given quantile."""
+            threshold = data.quantile(quantile)
+            return data.apply(lambda x: x > threshold)
+
+        def find_smallest_and_largest(data, quantile):
+            """ Return boolean vector of data points outside of the given quantile."""
+            rest = 1.0 - quantile
+            quantiles = [rest/2.0, 1.0 - rest/2.0]
+            thresholds = list(data.quantile(quantiles))
+            return data.apply(lambda x: x < thresholds[0] or x > thresholds[1])
+
+        def find_smallest_and_largest_asym(data, quantile):
+            """ Return boolen vector of data to remove such that quantile/2
+                is kept in both non-negative and non-positive subsets
+                of data."""
+            rest = 1.0 - quantile
+
+            neg_threshold = data[data <= 0.0].quantile(rest/2.0)
+            pos_threshold = data[data >= 0.0].quantile(1.0 - rest/2.0)
+
```

Comment:
`neg_threshodl` should be based on `<` instead of `<=`

And maybe `pos` should be renamed `non_negative`

Reply:
I'd like to keep `<=` and not to rename the variables :~) There is indeed a slight disagreement between the names and the semantics, but I think that introducing smth like `non_negative` will result in an overhead.

Reply:
In a dataset with zeroes, `<=` will result in too many values being removed

While there might be a lot of debate about which values to remove, we should be consistent on how many values are removed. I.e. That percentile _p_ will result in _(1-p)*n_ values being removed. This requires that we have non-overlapping predicates here

(Although, I admit that I still don't know what to do with NaNs)

Reply:
Aha. It seems like we're moving towards the following rule, based on the `skipna` thing in `_choose_threshold_type`:

"Given `n` datapoints - of which `f` are finite (i.e. neither `nan` nor `(-)inf`) - and a quantile `q`, we will remove `f*(1-q)` of them (rounded down the the nearest integer). The threshold_type will control which ones to remove, but won't affect how many are removed"

Is that correct?

Reply:
For upper/lower/two-sides -- yes.

For two-sides-asym it applies to both non-negative and non-positive ones separately with q' = q/2.

Reply:
So in the extreme case where everything is zero (we have some KPIs which have lots of zeroes in them), we will remove `2*f*(1-q)` of them

Reply:
> ... desired behaviour?

One issue is that I don't really know what I think the desired behaviour is. Maybe a Speclog would help here

In fact, I think I'll make such a document right now, and ask for feedback on it.

Reply:
... although, on second thoughts, I won't take that initiative right now. Let's ask around

Reply:
In the updated code line, line 330 reads
```
neg_threshold = data[data <  0.0].quantile(rest/2.0)
```

Reply:
Thanks for changing that. Now, with `data<0` and `data>=0`, we have non-overlapping sets used for defining the thresholds